File: src/GrammaticOptimization/Grammar.java
/*
 * Grammar.java
 *
 * Created on July 26, 2007, 6:09 PM
 *
 * To change this template, choose Tools | Template Manager
 * and open the template in the editor.
 */

package GrammaticOptimization;

/**
 *
 * @author adilraja
 */
import java.util.Iterator;

public abstract class Grammar extends Mapper{
    private boolean validGrammar;
    private int startSymbol;
    
    protected abstract boolean genotype2phenotype();
    protected abstract boolean phenotype2genotype();
    
    /**
     * Creates a new instance of Grammar 
     */
    
    public Grammar() {
        super();
        setValidGrammar(false);
	startSymbol=0;
    }
    
    /**
     *Constructor setting the genotype structure of this mapper to newGenotype.
     */
    public Grammar(final Genotype newGenotype){
        super(newGenotype);
        setValidGrammar(false);
	startSymbol=0;
    }
    
    
    /**
     *Constructor setting the phenotype structure of this mapper to newPhenotype.
     */
    public Grammar(final Phenotype newPhenotype){
        super(newPhenotype);
        setValidGrammar(false);
	startSymbol=0;
    }
    
    /**
     *Copy Constructor
     */
    public Grammar(final Grammar copy){
        super();
        setValidGrammar(copy.getValidGrammar());
	startSymbol=copy.startSymbol;
    }
    
    /**
     * Return the validity of the current grammar.
     */
    
    
    public final boolean getValidGrammar(){
	return validGrammar;
    }

    /**
     *Set the validity of the grammar.
     */
    protected void setValidGrammar(final boolean newValidGrammar){
	validGrammar=newValidGrammar;
    }
    
/**
 *Return current start symbol.
 */
    
public final Symbol getStartSymbol(){
	//return front().lhs.front();
	//return (*this)[startSymbol].lhs.front();
        Symbol symb=null;
        if (startSymbol<this.size())
        {
         symb=this.get(startSymbol).lhs.get(0);
        }
        return symb;
}

/**
 *Change start symbol by index on Vector of rules.
 */
public boolean setStartSymbol(final int index){
	if(index<this.size()){// Check boundaries.
		startSymbol=index;
		genotype2phenotype();// Update phenotype.
		return true;
	}
	return false;
}


/**
 *Change start symbol by symbol pointer.
 */
public boolean setStartSymbol(final Symbol newStartSymbol){
	int ii=0;
	Iterator<Rule> it=this.iterator();
	while(it.hasNext()){
		// Work by *pointer*.
		if(it.next().lhs.get(0).getSymbol().compareTo(newStartSymbol.getSymbol())==0){
			startSymbol=ii;
			genotype2phenotype();// Update phenotype.
			return true;
		}
		ii++;
	}
	return false;
}

/**
 *Change start symbol by string.
 */

public boolean setStartSymbol(final String newStartSymbol){
	int ii=0;
	Iterator<Rule> it=this.iterator();
	while(it.hasNext()){
		// Work by *string*.
		String currentStartSymbol=it.next().lhs.get(0).getSymbol();
		if(currentStartSymbol.compareTo(newStartSymbol)==0){
			startSymbol=ii;
			genotype2phenotype();// Update phenotype.
			return true;
		}
		ii++;
	}
	return false;
}

/**
 *Return pointer to current start rule.
 */

public final Rule getStartRule(){
	return this.get(0);
}

}
//complete

File: libGEC++/libGE-0.26/src/Grammar.cpp
// Grammar.cpp -*- C++ -*-
#ifndef _GRAMMAR_CPP_
#define _GRAMMAR_CPP_

//#include <config.h>

#include<iostream>

#include "Grammar.h"

///////////////////////////////////////////////////////////////////////////////
// Default constructor.
Grammar::Grammar():Mapper(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Grammar::Grammar()' called\n";
#endif
	setValidGrammar(false);
	startSymbol=0;
}

///////////////////////////////////////////////////////////////////////////////
// Constructor setting the genotype structure of this mapper to newGenotype.
Grammar::Grammar(const Genotype &newGenotype):Mapper(newGenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Grammar::Grammar(const Genotype &,unsigned int=0)' called\n";
#endif
	setValidGrammar(false);
	startSymbol=0;
}

///////////////////////////////////////////////////////////////////////////////
// Constructor setting the phenotype structure of this mapper to newPhenotype.
Grammar::Grammar(const Phenotype &newPhenotype):Mapper(newPhenotype){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Grammar::Grammar(const Phenotype &,unsigned int=0)' called\n";
#endif
	setValidGrammar(false);
	startSymbol=0;
}

///////////////////////////////////////////////////////////////////////////////
// Copy constructor.
Grammar::Grammar(const Grammar& copy):Mapper(copy),vector<Rule>(copy){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Grammar::Grammar(const Grammar&)' called\n";
#endif
	setValidGrammar(copy.getValidGrammar());
	startSymbol=copy.startSymbol;
}

///////////////////////////////////////////////////////////////////////////////
// Destructor: delete all Productions first, and then all rules.
Grammar::~Grammar(){
#if (DEBUG_LEVEL >= 2)
	cerr << "'Grammar::~Grammar()' called\n";
#endif
	iterator ruleIt=begin();
	while(ruleIt!=end()){
		(*ruleIt).clear();
		ruleIt++;
	}
	clear();
}

///////////////////////////////////////////////////////////////////////////////
// Return the validity of the current grammar.
bool Grammar::getValidGrammar() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Grammar::getValidGrammar() const' called\n";
#endif
	return validGrammar;
}

///////////////////////////////////////////////////////////////////////////////
// Set the validity of the grammar.
void Grammar::setValidGrammar(const bool newValidGrammar){
#if (DEBUG_LEVEL >= 2)
	cerr << "'void Grammar::setValidGrammar(const bool)' called\n";
#endif
	validGrammar=newValidGrammar;
}

///////////////////////////////////////////////////////////////////////////////
// Return pointer to current start symbol.
const Symbol* Grammar::getStartSymbol() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'const Symbol* Grammar::getStartSymbol() const' called\n";
#endif
	//return front().lhs.front();
	return (*this)[startSymbol].lhs.front();
}

///////////////////////////////////////////////////////////////////////////////
// Change start symbol by index on Vector of rules.
bool Grammar::setStartSymbol(const unsigned int index){
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Grammar::setStartSymbol(const unsigned int)' called\n";
#endif
	if(index<size()){// Check boundaries.
		startSymbol=index;
		genotype2Phenotype();// Update phenotype.
		return true;
	}
	return false;
}

///////////////////////////////////////////////////////////////////////////////
// Change start symbol by symbol pointer.
bool Grammar::setStartSymbol(const Symbol *newStartSymbol){
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Grammar::setStartSymbol(const Symbol*)' called\n";
#endif
	int ii=0;
	iterator it=begin();
	while(it!=end()){
		// Work by *pointer*.
		if(it->lhs.front()==newStartSymbol){
			startSymbol=ii;
			genotype2Phenotype();// Update phenotype.
			return true;
		}
		it++;ii++;
	}
	return false;
}

///////////////////////////////////////////////////////////////////////////////
// Change start symbol by string.
bool Grammar::setStartSymbol(const string &newStartSymbol){
#if (DEBUG_LEVEL >= 2)
	cerr << "'bool Grammar::setStartSymbol(const string&)' called\n";
#endif
	int ii=0;
	iterator it=begin();
	while(it!=end()){
		// Work by *string*.
		string currentStartSymbol=*(it->lhs.front());
		if(currentStartSymbol==newStartSymbol){
			startSymbol=ii;
			genotype2Phenotype();// Update phenotype.
			return true;
		}
		it++;ii++;
	}
	return false;
}

///////////////////////////////////////////////////////////////////////////////
// Return pointer to current start rule.
const Rule* Grammar::getStartRule() const{
#if (DEBUG_LEVEL >= 2)
	cerr << "'const Rule* Grammar::getStartRule() const' called\n";
#endif
	return &front();
}

#endif

